src/CanisterAssets/lib.mo

```diff
@@ -19,26 +20,76 @@ module {
 
     var _assets = Buffer.Buffer<Types.Asset>(0);
 
-    public func toStableChunk(chunkSize : Nat, chunkIndex : Nat) : Types.StableChunk {
-      ?#v1({
-        assets = Buffer.toArray(_assets);
-      });
+    let _bytesPerChunk = 500_000; // 500kb
+    var _biggestAssetSize = 10_000; // 10kb
+
+    public func getChunkCount() : Nat {
+      var count = _assets.size() * _biggestAssetSize / _bytesPerChunk;
+      if (_assets.size() * _biggestAssetSize % _bytesPerChunk != 0) {
+        count += 1;
+      };
+      return Nat.max(1, count);
+    };
+
+    public func toStableChunk(chunkSize_ignored : Nat, chunkIndex : Nat) : Types.StableChunk {
+      let chunkSize = _bytesPerChunk / _biggestAssetSize;
+      let start = Nat.min(_assets.size(), chunkSize * chunkIndex);
+      let count = Nat.min(chunkSize, _assets.size() - start);
+      let assetsChunk = if (_assets.size() == 0 or count == 0) {
+        []
+      }
+      else {
+        Buffer.toArray(Buffer.subBuffer(_assets, start, count));
+      };
+
+      if (chunkIndex == 0) {
+        return ?#v1({
+          assetsCount = _assets.size();
+          assetsChunk;
+        });
+      }
+      else if (chunkIndex < getChunkCount()) {
+        return ?#v1_chunk({ assetsChunk });
+      }
+      else {
+        null;
+      };
     };
 
     public func loadStableChunk(chunk : Types.StableChunk) {
       switch (chunk) {
         // TODO: remove after upgrade vvv
         case (?#legacy(state)) {
           _assets := Buffer.fromArray(state._assetsState);
+          _updateBiggestAssetSize();
         };
         // TODO: remove after upgrade ^^^
         case (?#v1(data)) {
-          _assets := Buffer.fromArray(data.assets);
+          _assets := Buffer.Buffer<Types.Asset>(data.assetsCount);
+          _assets.append(Buffer.fromArray(data.assetsChunk));
+          _updateBiggestAssetSize();
+        };
+        case (?#v1_chunk(data)) {
+          _assets.append(Buffer.fromArray(data.assetsChunk));
+          _updateBiggestAssetSize();
         };
         case (null) {};
       };
     };
 
+    func _updateBiggestAssetSize() {
```

Comment:
instead of going over the entire assets, can we instead only look at the current chunk to see if any asset is bigger than the current value?

Reply:
fixed